src/abstractarray.jl

```diff
@@ -39,169 +37,88 @@ Base.linearindexing{T<:StaticArray}(::Type{T}) = Base.LinearFast()
-            SArray{S, $(eltype(SA)), $(length(S)), $(prod(S))}
-        end
-    end
-end
+mutable_similar_type{T,S}(::Type{T}, s::Size{S}, ::Type{Val{0}}) = SizedArray{(),T,0,0}
+@generated mutable_similar_type{T,S}(::Type{T}, s::Size{S}, ::Type{Val{1}}) = MVector{S[1],T}
+@generated mutable_similar_type{T,S}(::Type{T}, s::Size{S}, ::Type{Val{2}}) = MMatrix{S[1],S[2],T,prod(S)}
+@generated mutable_similar_type{T,S,D}(::Type{T}, s::Size{S}, ::Type{Val{D}}) = MArray{S,T,D,prod(S)}
 
-# Some specializations for the mutable case
-@pure similar_type{MA<:Union{MVector,MMatrix,MArray,SizedArray}}(::Union{MA,Type{MA}}, size::Int) = MVector{size, eltype(MA)}
-@pure similar_type{MA<:Union{MVector,MMatrix,MArray,SizedArray}}(::Union{MA,Type{MA}}, sizes::Tuple{Int}) = MVector{sizes[1], eltype(MA)}
+# `SizedArray` stays the same, and also takes over an `Array`.
+similar_type{SA<:SizedArray,T,S}(::Type{SA},::Type{T},s::Size{S}) = sizedarray_similar_type(T,s,length_val(s))
+similar_type{A<:Array,T,S}(::Type{A},::Type{T},s::Size{S}) = sizedarray_similar_type(T,s,length_val(s))
 
-@pure similar_type{MA<:Union{MVector,MMatrix,MArray,SizedArray}}(::Union{MA,Type{MA}}, sizes::Tuple{Int,Int}) = MMatrix{sizes[1], sizes[2], eltype(MA), sizes[1]*sizes[2]}
+sizedarray_similar_type{T,S,D}(::Type{T},s::Size{S},::Type{Val{D}}) = SizedArray{S,T,D,D}
 
-@pure similar_type{MA<:Union{MVector,MMatrix,MArray,SizedArray},N}(::Union{MA,Type{MA}}, sizes::Tuple{Vararg{Int,N}}) = MArray{sizes, eltype(MA), N, prod(sizes)}
+# Field vectors are user controlled, and default to SVector, etc
 
-@generated function similar_type{MA<:Union{MVector,MMatrix,MArray,SizedArray},S}(::Union{MA,Type{MA}}, ::Size{S})
-    if length(S) == 1
-        return quote
-            $(Expr(:meta, :inline))
-            MVector{$(S[1]), $(eltype(MA))}
-        end
-    elseif length(S) == 2
-        return quote
-            $(Expr(:meta, :inline))
-            MMatrix{$(S[1]), $(S[2]), $(eltype(MA))}
-        end
-    else
-        return quote
-            $(Expr(:meta, :inline))
-            MArray{S, $(eltype(MA)), $(length(S)), $(prod(S))}
-        end
-    end
-end
+"""
+    similar(static_array)
+    similar(static_array, T)
+    similar(array, ::Size)
+    similar(array, T, ::Size)
+
+Constructs and returns a mutable but statically-sized array (i.e. a `StaticArray`). If the
+input `array` is not a `StaticArray`, then the `Size` is required to determine the output
+size (or else a dynamically sized array will be returned).
+"""
+similar{SA<:StaticArray}(::SA) = similar(SA,eltype(SA))
+similar{SA<:StaticArray}(::Type{SA}) = similar(SA,eltype(SA))
 
-# And also similar() returning mutable StaticArrays
-@inline similar{SV <: StaticVector}(::SV) = MVector{length(SV),eltype(SV)}()
-@inline similar{SV <: StaticVector, T}(::SV, ::Type{T}) = MVector{length(SV),T}()
+similar{SA<:StaticArray,T}(::SA,::Type{T}) = similar(SA,T,Size(SA))
+similar{SA<:StaticArray,T}(::Type{SA},::Type{T}) = similar(SA,T,Size(SA))
```

Comment:
So now we've got `similar` based on input `StaticArray` type as well as value.  I guess that makes sense as they all still return a value which can be written into. Did did we have it before anywhere?